### PR TITLE
Add RecordDeclaration to IsNamedTypeDeclarationBlock

### PR DESCRIPTION
**Pull request by Youssef1313**
I'm not sure if this changes behavior or not.

### CODE REVIEW

File: src/Microsoft.CodeAnalysis.Analyzers/CSharp/MetaAnalyzers/CSharpDiagnosticAnalyzerAPIUsageAnalyzer.cs [hunk 1]
```diff
@@ -14,6 +14,7 @@ protected override bool IsNamedTypeDeclarationBlock(SyntaxNode syntax)
             switch (syntax.Kind())
             {
                 case SyntaxKind.ClassDeclaration:
+                case SyntaxKind.RecordDeclaration:
```
Comment (Evangelink) [inline src/Microsoft.CodeAnalysis.Analyzers/CSharp/MetaAnalyzers/CSharpDiagnosticAnalyzerAPIUsageAnalyzer.cs:17]: Let's add a test for this.

**Comment by Evangelink** [comment]
@Youssef1313 Build is failing because the syntax is not found on all compilations. You might need to have this case handled only for Roslyn 3+ (see https://github.com/dotnet/roslyn-analyzers/pull/4223/files#diff-d8afcf96221c79be0c7c1d6e2defd475R754)

File: src/Microsoft.CodeAnalysis.Analyzers/CSharp/MetaAnalyzers/CSharpDiagnosticAnalyzerAPIUsageAnalyzer.cs (hunk 1 above)
Comment (Youssef1313) [inline src/Microsoft.CodeAnalysis.Analyzers/CSharp/MetaAnalyzers/CSharpDiagnosticAnalyzerAPIUsageAnalyzer.cs:17]:
```suggestion
 #if CODEANALYSIS_V3_OR_BETTER
                case SyntaxKind.RecordDeclaration:
#endif
```

Comment (Youssef1313) [reply src/Microsoft.CodeAnalysis.Analyzers/CSharp/MetaAnalyzers/CSharpDiagnosticAnalyzerAPIUsageAnalyzer.cs:18]: @Evangelink I'm not sure how the test should look like to have a different behavior before and after.

Comment (Evangelink) [reply src/Microsoft.CodeAnalysis.Analyzers/CSharp/MetaAnalyzers/CSharpDiagnosticAnalyzerAPIUsageAnalyzer.cs:18]: I haven't tested but you might be able to copy a test raising an issue and change class to be record and show that this does raise an issue.

Comment (Youssef1313) [reply src/Microsoft.CodeAnalysis.Analyzers/CSharp/MetaAnalyzers/CSharpDiagnosticAnalyzerAPIUsageAnalyzer.cs:18]: @Evangelink tests are based on classes that inherit DiagnosticAnalyzer. Records on the other hand cannot inherit classes.
I'll debug this to make sure if records can come across this code path (e.g. record nested inside a class that inherit diagnostic analyzer)

File: src/Microsoft.CodeAnalysis.Analyzers/CSharp/MetaAnalyzers/CSharpDiagnosticAnalyzerAPIUsageAnalyzer.cs [hunk 2]
```diff
@@ -14,6 +14,9 @@ protected override bool IsNamedTypeDeclarationBlock(SyntaxNode syntax)
             switch (syntax.Kind())
             {
                 case SyntaxKind.ClassDeclaration:
+ #if CODEANALYSIS_V3_OR_BETTER
```
Comment (Youssef1313) [inline src/Microsoft.CodeAnalysis.Analyzers/CSharp/MetaAnalyzers/CSharpDiagnosticAnalyzerAPIUsageAnalyzer.cs:17]:
```suggestion
#if CODEANALYSIS_V3_OR_BETTER
```

File: src/Microsoft.CodeAnalysis.Analyzers/CSharp/MetaAnalyzers/CSharpDiagnosticAnalyzerAPIUsageAnalyzer.cs (hunk 1 above)
Comment (Evangelink) [reply src/Microsoft.CodeAnalysis.Analyzers/CSharp/MetaAnalyzers/CSharpDiagnosticAnalyzerAPIUsageAnalyzer.cs:18]: Oh I was left on an old proposition for the record where the inheritance was allowed but it seems it is no longer the case (https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/records). So basically what's the purpose of your change of there is no way to reach this point?

Comment (Youssef1313) [reply src/Microsoft.CodeAnalysis.Analyzers/CSharp/MetaAnalyzers/CSharpDiagnosticAnalyzerAPIUsageAnalyzer.cs:18]: @Evangelink I didn't mean this is unreachable with my previous comment. (I also didn't mean this is reachable).

Until now I'm not sure whether this is reachable or not.
And if it's reachable, will this have a behavior change, or just returning early for some cases.

The method is called in two different places in the analyzer, which should have two different purposes. But the current implementation has struct, enum, interface. If all of these are reachable, then records shouldn't different.

I'll debug soon to get feel of how the code flows and get an understanding of it.

**Comment by mavasani** [comment]
@Youssef1313 are you able to add a regression test for this?

**Comment by Youssef1313** [comment]
@mavasani Nope, I *think* this helps to bail-out early for irrelevant nodes.

**Pull request merged**